### ISSUE TITLE
Bug in LSTMBlockCellBpropWithCUDA when using peephole connections

### ISSUE THREAD

**Issue by chaaland** [tensorflow/tensorflow#34171]
<em>Please make sure that this is a bug. As per our [GitHub Policy](https://github.com/tensorflow/tensorflow/blob/master/ISSUES.md), we only address code/doc bugs, performance issues, feature requests and build/installation issues on GitHub. tag:bug_template</em>

**System information**
- Have I written custom code (as opposed to using a stock example script provided in TensorFlow):
No

- OS Platform and Distribution (e.g., Linux Ubuntu 16.04):
Windows 10

- Mobile device (e.g. iPhone 8, Pixel 2, Samsung Galaxy) if the issue happens on mobile device:

- TensorFlow installed from (source or binary):
Source
- TensorFlow version (use command below):
r2.0
- Python version:
3.6.6
- Bazel version (if compiling from source):
- GCC/Compiler version (if compiling from source):
- CUDA/cuDNN version:
9.0/7.4
- GPU model and memory:
Quadro M2200

**Describe the current behavior**
In the backward propagation of the LSTM GPU code, there's a [broadcast operation included in the gradient computation](https://github.com/tensorflow/tensorflow/blob/722b96b22926dbc05881c35cb63fd342c6843112/tensorflow/core/kernels/rnn/lstm_ops_gpu.cu.cc#L398) of `cs_prev_grad` involving `wci` and `wcf`
```c++
  if (use_peephole) {
    Eigen::array<Eigen::DenseIndex, 2> p_shape({1, cell_size});
    Eigen::array<Eigen::DenseIndex, 2> p_broadcast_shape({batch_size, 1});
    cs_prev_grad.device(d) =
        cs_prev_grad + di * wci.reshape(p_shape).broadcast(p_broadcast_shape) +
        df * wcf.reshape(p_shape).broadcast(p_broadcast_shape);
    wci_grad.device(d) = (di * cs_prev).sum(Eigen::array<int, 1>({0}));
    wcf_grad.device(d) = (df * cs_prev).sum(Eigen::array<int, 1>({0}));
    wco_grad.device(d) = (do_ * cs).sum(Eigen::array<int, 1>({0}));
  }
}
```
Just before this code, the [GPU kernel](https://github.com/tensorflow/tensorflow/blob/722b96b22926dbc05881c35cb63fd342c6843112/tensorflow/core/kernels/rnn/lstm_ops_gpu.cu.cc#L358) `lstm_gates_bprop` is launched which calculates the gradient of `cs_prev_grad` using `wcf` and `wci`

```c++
__global__ void lstm_gates_bprop(
    const T* cs_prev,  // [batch_size, cell_size]
    const T* h_prev,   // [batch_size, cell_size]
    const T* w,        // [input_size + cell_size, 4 * cell_size]
    const T* wci,      // [cell_size]
    const T* wcf,      // [cell_size]
    const T* wco,      // [cell_size]
    const T* b,        // [4 * cell_size]
    const T* i,        // [batch_size, cell_size]
    const T* cs,       // [batch_size, cell_size]
    const T* f,        // [batch_size, cell_size]
    const T* o,        // [batch_size, cell_size]
    const T* ci,       // [batch_size, cell_size]
    const T* co,       // [batch_size, cell_size]
    const T* cs_grad,  // [batch_size, cell_size]
    const T* h_grad,   // [batch_size, cell_size]
    T* do_,            // [batch_size, cell_size]
    T* dcs,            // [batch_size, cell_size]
    T* dci,            // [batch_size, cell_size]
    T* df,             // [batch_size, cell_size]
    T* di,             // [batch_size, cell_size]
    T* dgates,         // [input_size + cell_size, 4 * cell_size]
    T* cs_prev_grad,   // [batch_size, cell_size]
    const int batch_size, const int cell_size, const bool use_peephole) {
  const int batch_id = blockIdx.x * blockDim.x + threadIdx.x;
  const int act_id = blockIdx.y * blockDim.y + threadIdx.y;

  if (batch_id >= batch_size || act_id >= cell_size) return;

  const int gid = batch_id * cell_size * 4 + act_id;
  const int cid = batch_id * cell_size + act_id;

  // bunch of code

  dgates[gid + 0 * cell_size] = di_local;
  dgates[gate_c_offset(gate_layout, cell_size)] = dci_local;
  dgates[gate_f_offset(gate_layout, cell_size)] = df_local;
  dgates[gid + 3 * cell_size] = do_local;

  cs_prev_grad[cid] = dcs_local * f_local;
  if (use_peephole) {
    cs_prev_grad[cid] += di_local * wci[act_id] + df_local * wcf[act_id];
  }
```

**Describe the expected behavior**
Unless I misunderstand, the gradient when using peephole connections is incorrect. The peephole gradient should be taken into account _either_ in the GPU kernel _or_ by the eigen computation, but not both.

**Code to reproduce the issue**
Provide a reproducible test case that is the bare minimum necessary to generate the problem.

**Other info / logs**
Include any logs or source code that would be helpful to diagnose the problem. If including tracebacks, please include the full traceback. Large logs and files should be attached.

**Comment by qlzh727** [comment]
Adding original author @ekelsen

**Comment by qlzh727** [comment]
The original change is added in https://github.com/tensorflow/tensorflow/commit/61b521679b9468b4b4ffa792a1ebc6a5ecd28317.

And also add true author @duckworthd

**Comment by chaaland** [comment]
Additionally, I also think these kernels have some slowness due to not accessing coalesced memory. The kernel launch indicates `batch_size` is being thought of as the x dimension (cartesian) when it is actually the row dimension of the matrix (CUDA uses image notation rather than matrix notation.

Looking at the (definition of the cid variable)[https://github.com/tensorflow/tensorflow/blob/722b96b22926dbc05881c35cb63fd342c6843112/tensorflow/core/kernels/rnn/lstm_ops_gpu.cu.cc#L312] it uses `act_id` as the offset. Yet `act_id` is defined in terms of the `threadIdx.y` which is the slower varying dimension. We should prefer it being dependent on `threadIdx.x` for coalesced memory access

**Comment by kumariko** [comment]
Could you please let us know if this issue still persists ? If it it resolved then please feel free to move this issue to close status ? please check **[link](https://www.tensorflow.org/api_docs/python/tf/keras/experimental/PeepholeLSTMCell)** . Thanks!

**Comment by chaaland** [comment]
It still looks like the bug is there having just looked at the code now

**Comment by tilakrayal** [comment]
@chaaland,
The required changes were made in the **lstm_ops_gpu.cu.cc** file and also tf.keras.experimental.PeepholeLSTMCell was deprecated in the latest version.

https://github.com/tensorflow/tensorflow/blob/61b521679b9468b4b4ffa792a1ebc6a5ecd28317/tensorflow/contrib/rnn/kernels/lstm_ops_gpu.cu.cc
```
               LSTMBlockCellBpropWithCUDA<T>(                                             \
        ctx, d, x, cs_prev, h_prev, w, wci, wcf, wco, b, i, cs, f, o, ci, co,  \
        cs_grad, h_grad, do_, dcs, dci, df, di, dicfo, cs_prev_grad, wci_grad, \
        wcf_grad, wco_grad, batch_size_, cell_size_, use_peephole);  
```

**Comment by tilakrayal** [comment]
Closing this as stale. Please reopen if this is still a valid request. Thank you!
